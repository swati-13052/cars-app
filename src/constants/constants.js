export const CAR_FILTERS = {
    color: 'color',
    manufacturer: 'manufacturer'
}