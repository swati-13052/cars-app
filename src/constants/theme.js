import { createTheme } from "@mui/material";
import { createContext } from "react";

export const theme = {
    COLOURS: {
        GRAY_LIGHT: '#EDEDED',
        GRAY_DARK: '#4A4A4A',
        ORANGE_LIGHT: '#EA7F28',
        ORANGE_DARK: '#D37324'
    },
    SPACING: {
        SMALL: '8px',
        MEDIUM: '12px',
        LARGE: '32px',
    }
}

export const ThemeContext = createContext();

export const ButtonTheme = createTheme({
    palette: {
        primary: {
            light: '#ff7961',
            main: '#EA7F28',
            dark: '#D37324',
            contrastText: '#fff'
        },
        action: {
            hover: '#D37324'
        }
    }
})