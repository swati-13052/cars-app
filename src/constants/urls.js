import { payloadToQueryParams } from './../utils/util';

export const getAllCarsURL = (queryParams) => `https://auto1-mock-server.herokuapp.com/api/cars${payloadToQueryParams(queryParams)}`;
export const getAllColoursURL = `https://auto1-mock-server.herokuapp.com/api/colors`;
export const getAllManufacturerURL = `https://auto1-mock-server.herokuapp.com/api/manufacturers`;
export const getCarDetails = (stockNumber) => `https://auto1-mock-server.herokuapp.com/api/cars/${stockNumber}`;
export const logoURL = `https://auto1-js-task-api--mufasa71.repl.co/images/logo.png`;