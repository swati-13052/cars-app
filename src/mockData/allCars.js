export const allCars = [
    {
        "stockNumber": 30257,
        "manufacturerName": "Skoda",
        "modelName": "Forman",
        "color": "green",
        "mileage": {
            "number": 100099,
            "unit": "km"
        },
        "fuelType": "Petrol",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
    }

]