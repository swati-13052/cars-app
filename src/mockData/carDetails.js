export const carDetails = {
    "car": {
      "stockNumber": 10035,
      "manufacturerName": "Mercedes-Benz",
      "modelName": "GL-Klasse",
      "color": "blue",
      "mileage": {
        "number": 180222,
        "unit": "km"
      },
      "fuelType": "Diesel",
      "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
    }
  }