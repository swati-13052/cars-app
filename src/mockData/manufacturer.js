export const manufacturers = [
    {
        "name": "Audi",
        "models": [
            {
                "name": "100"
            },
            {
                "name": "200"
            },
            {
                "name": "50"
            },
            {
                "name": "80/90"
            },
            {
                "name": "A1"
            },
            {
                "name": "A2"
            },
            {
                "name": "A3"
            },
            {
                "name": "A4"
            },
            {
                "name": "A4 Allroad"
            },
            {
                "name": "A5"
            },
            {
                "name": "A6"
            },
            {
                "name": "A6 Allroad"
            },
            {
                "name": "A7"
            },
            {
                "name": "A8"
            },
            {
                "name": "Allroad"
            },
            {
                "name": "Cabriolet"
            },
            {
                "name": "Coupe"
            },
            {
                "name": "Q2"
            },
            {
                "name": "Q3"
            },
            {
                "name": "Q5"
            },
            {
                "name": "Q7"
            },
            {
                "name": "Quattro"
            },
            {
                "name": "R8"
            },
            {
                "name": "RS 3"
            },
            {
                "name": "RS 4"
            },
            {
                "name": "RS 5"
            },
            {
                "name": "RS 6"
            },
            {
                "name": "RS 7"
            },
            {
                "name": "RS Q3"
            },
            {
                "name": "S1"
            },
            {
                "name": "S3"
            },
            {
                "name": "S4"
            },
            {
                "name": "S5"
            },
            {
                "name": "S6"
            },
            {
                "name": "S7"
            },
            {
                "name": "S8"
            },
            {
                "name": "SQ5"
            },
            {
                "name": "SQ7"
            },
            {
                "name": "TT"
            },
            {
                "name": "TT Roadster"
            },
            {
                "name": "V8"
            }
        ]
    }
]