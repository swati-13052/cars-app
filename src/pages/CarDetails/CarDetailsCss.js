import styled from "styled-components";
import Button from '@mui/material/Button';

export const CarInfoContainer = styled.div`
    width: 70%;
    margin: auto;
    margin-top: 24px;
    display: flex;
    justify-content: space-between;
    min-height: 400px;
`

export const CarInfo = styled.div`{
 width: 60%;
    .car-name {
        font-size: 32px;
        font-weight: 700;
        color: ${props => props.theme.COLOURS.GRAY_DARK};
        margin-bottom: 24px;
    }
    .car-info {
        font-size: 24px;
        margin-bottom: 24px;
        color: ${props => props.theme.COLOURS.GRAY_DARK};
        font-weight: 500;
    }
    .car-availability-text {
        color: ${props => props.theme.COLOURS.GRAY_DARK};
        font-weight: 500;
        line-height: 24px;
    }
 
}`

export const AddButton = styled(Button)`
    &&& {
        margin-top: 16px;
        text-align: right;
        width:50%;
    }
`


export const AddCarBox = styled.div`
    border: 1px solid ${props => props.pagetheme.COLOURS.GRAY_LIGHT};
    width: 30%;
    padding: 24px;
    max-height: 140px;
    .add-car-text {
        line-height: 24px;
        font-weight: 500;
        color: ${props => props.pagetheme.COLOURS.GRAY_DARK};
    }
`

export const CarImage = styled.div`
    width: 100%;
    height: 400px;
    display: flex;
    align-items: center;
    justify-content: center;
    background: ${props => props.theme.COLOURS.GRAY_LIGHT};

    .car-image {
        width: 200px;
        height: 200px;
    }
`

export const LoadingCarDetailsContainer = styled.div`
    display: flex;
    justify-content: center;
    min-height: 300px;
    align-items: center;
`

export const CarInfoShimmer = styled.div`
    display: flex;
    flex-direction: column;
`