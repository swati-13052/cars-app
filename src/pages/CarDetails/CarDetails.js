import { useContext, useEffect, useState } from "react";
import { useParams, useNavigate } from "react-router-dom";
import { ThemeContext } from "../../constants/theme";
import { carService } from "../../services/carService";
import { firstCharUpperCase } from "../../utils/util";
import { CarInfoContainer, CarInfo, AddButton, AddCarBox, CarImage } from './CarDetailsCss';
import CircularProgress from '@mui/material/CircularProgress';
import CarDetailsSkeleton from "./CarDetailsSkeleton";

function CarDetails() {

    const [carDetails, setCarDetails] = useState({});
    const { stockNumber } = useParams();
    const theme = useContext(ThemeContext);
    const [favouriteCollections, setFavouriteCollections] = useState([]);
    const [isAddingCar, setIsAddingCar] = useState(false);
    const [isLoadingCarDetails, setIsLoadingCarDetails] = useState(false);
    let navigate = useNavigate();

    useEffect(() => {
        setIsLoadingCarDetails(true);
        carService.getCarDetailsForStockNumber(stockNumber)
            .then(data => {
                setCarDetails(data?.car || {});
            })
            .catch(error => {
                if (error.message === 404) {
                    navigate('*')
                }
            })
            .finally(() => {
                setIsLoadingCarDetails(false);
            })

        let addedCars = localStorage.getItem('favouriteCollections');
        if (addedCars) {
            setFavouriteCollections(JSON.parse(addedCars));
        }
    }, [stockNumber, navigate]);

    useEffect(() => {
        localStorage.setItem('favouriteCollections', JSON.stringify(favouriteCollections));
    }, [favouriteCollections]);

    const getCarInfo = () => {
        if (Object.keys(carDetails).length) {
            const carInfo = [`Stock # ${carDetails?.stockNumber}`, `${carDetails?.mileage?.number} ${carDetails?.mileage?.unit}`, `${carDetails?.fuelType}`, `${firstCharUpperCase(carDetails?.color || '')}`];
            return carInfo.join(" - ");
        } else {
            return '-'
        }
    };

    const handleSaveCar = () => {
        setIsAddingCar(true);
        setTimeout(() => {
            let carsAdded = [...favouriteCollections];
            const carIndexInFavourites = carsAdded.findIndex(car => car.stockNumber === carDetails.stockNumber);
            if (carIndexInFavourites >= 0) {
                carsAdded.splice(carIndexInFavourites, 1);
            } else {
                carsAdded = [...carsAdded, carDetails];
            }
            setFavouriteCollections(carsAdded);
            setIsAddingCar(false);
        }, 2000)

    }

    const isCarAddedToFavourite = () => {
        const carIndexInFavourites = favouriteCollections.findIndex(car => car.stockNumber === carDetails?.stockNumber);
        return carIndexInFavourites >= 0 ? 'Remove' : 'Save';
    }

    return (
        <div>
            <CarImage theme={theme}>
                <img className="car-image" src={carDetails?.pictureUrl} alt="Loading.." />
            </CarImage>

            {isLoadingCarDetails ?
                <CarDetailsSkeleton/>
                :
                <CarInfoContainer>
                    <CarInfo theme={theme}>
                        <div data-testid="car-name-detail-page" className="car-name">{carDetails?.manufacturerName} {carDetails?.modelName}</div>
                        <div data-testid="car-info" className="car-info">{getCarInfo()}</div>
                        <div className="car-availability-text">This car is currently available and can be delivereds soon as tomorrow morning. Please be aware that delivery imes shown n this page are not definitive and may change ue to bad weather conditions</div>
                    </CarInfo>
                    <AddCarBox pagetheme={theme}>
                        <div className="add-car-text">If you like this car, click the button and save it in your collection of favourite items</div>
                        <AddButton disabled={isAddingCar} color="primary" onClick={() => handleSaveCar()} pagetheme={theme} variant="contained">
                            {isAddingCar && <CircularProgress size={20} />}
                            {!isAddingCar && isCarAddedToFavourite()}
                        </AddButton>
                    </AddCarBox>
                </CarInfoContainer>
            }
        </div>
    )
}

export default CarDetails;