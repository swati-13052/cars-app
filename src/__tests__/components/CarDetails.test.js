import React from 'react'
import { render, act } from '@testing-library/react';
import { theme, ThemeContext } from '../../constants/theme';
import { MemoryRouter } from 'react-router-dom';
import CarDetails from '../../pages/CarDetails/CarDetails';
import { carDetails } from '../../mockData/carDetails';

const renderWithContext = (
    component) => {
    return {
        ...render(
            <ThemeContext.Provider value={theme}>
                {component}
            </ThemeContext.Provider>, { wrapper: MemoryRouter })
    }
}

let originFetch;

beforeEach(() => {
    originFetch = global.fetch;
});

afterEach(() => {
    global.fetch = originFetch;
});

it('getCarDetails should have been called', async () => {
  const carDetailsResponse = carDetails;
  const mRes = { json: jest.fn().mockResolvedValueOnce(carDetailsResponse) };
  const mockedFetch = jest.fn().mockResolvedValueOnce(mRes);
  global.fetch = mockedFetch;
  await act( async () => renderWithContext(<CarDetails/>));
  expect(mockedFetch).toHaveBeenCalled();
});

it('should take a snapshot', () => {
    const { asFragment } = renderWithContext(<CarDetails />)
    expect(asFragment(<CarDetails />)).toMatchSnapshot()
});