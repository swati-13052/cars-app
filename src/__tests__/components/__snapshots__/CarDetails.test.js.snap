// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-eCYdqJ pWzAx"
    >
      <img
        alt="Loading.."
        class="car-image"
      />
    </div>
    <div>
      <div
        class="sc-bczRLJ ftEacv"
      >
        <div
          class="sc-gKXOVf hyfxAg"
        >
          <div
            class="sc-iBkjds kyxOsE"
            height="50px"
            width="400px"
          />
          <div
            class="sc-iBkjds hSBNiW"
            height="20px"
            width="500px"
          />
          <div
            class="sc-iBkjds hPqkyq"
            height="200px"
            width="500px"
          />
        </div>
        <div
          class="sc-hKMtZM gEpEHx"
        >
          <div
            class="sc-iBkjds bTQqDC"
            height="40px"
            width="300px"
          />
          <div
            class="sc-iBkjds dOfMWK"
            height="20px"
            width="100px"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
